### PR TITLE
More porting of vbruntime

### PR DESCRIPTION
**Pull request by cston**
The source is from https://github.com/Microsoft/referencesource/tree/master/Microsoft.VisualBasic/runtime/msvbalib.

There were minor cosmetic changes to comments and assert text.

The only changes to the implementation were:
- `LikeOperator.ExpandString()` does not handle full/half width mapping
- `VBBinder.InvokeMember()` does not handle `objType.IsCOMObject()`
- `LateBinding.GetCorrectIReflect()` does not handle `IsCOMObject()`.

### CODE REVIEW

**Comment by cston** [comment]
@333fred, @agocke, @danmosemsft, please review, thanks.

**Comment by cston** [comment]
cc @jaredpar, @KathleenDollard

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb [hunk 1]
```diff
@@ -0,0 +1,1253 @@
...
+                '         step 1:  can the Set succeed?
+                '         step 2:  perform the Set
+                '   the rvaluebase check would be done between 1 and 2
+                InternalLateSet(o, objType, name, args, paramnames, OptimisticSet, DefaultCallType)
+
+                If RValueBase AndAlso objType.IsValueType Then
+                    'note that objType is passed byref to InternalLateSet and that it 
+                    'should be valid by the time we get to this point
+                    Throw New Exception(GetResourceString(SR.RValueBaseForValueType, VBFriendlyName(objType, o), VBFriendlyName(objType, o)))
+                End If
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb]: Maybe remove some of the UNDONE, CONSIDER etc commentary

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LikeOperator.vb [hunk 2]
```diff
@@ -0,0 +1,1753 @@
...
+            Friend MaxSourceIndex As Integer
+            Friend CharCount As Integer
+
+            ' StringPatternStart, StringPatternEnd - there are the indices into the original source string
+            ' and are NOT indices into StringPattern.
+            '
+#If DEBUG Then
+            Private m_StringPatternStart As Integer  'For PatternType.[STRING]
+            Friend Property StringPatternStart() As Integer
+                Get
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LikeOperator.vb]: We can probably  lose all the  "!!!" In all this...

**Comment by danmosemsft** [comment]
Nice set of tests you ported.

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb [hunk 3]
```diff
@@ -0,0 +1,1235 @@
...
+            Dim binder As New VBBinder(Nothing)
+            binder.CacheMember(member)
+
+            Return objIReflect.InvokeMember(member.Name, VBLateBindingFlags Or flags, binder, target, args, Nothing, Nothing, Nothing)
+
+        End Function
+
+        Private Shared Function GetCorrectIReflect(ByVal o As Object, ByVal objType As Type) As IReflect
+
+            ' For a System.Type Object, we always use the underlying System.Type's IReflect implementation, because a System.Type's Implementation
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/LateBinding.vb]: This comment is about the removed `IReflect` section. Consider removing.

File: src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        }
+
+        [Theory]
+        [MemberData(nameof(FromString_Overflow_TestData))]
+        public void FromString_ThrowsOverflowException(string value)
+        {
+            Assert.Throws<OverflowException>(() => ByteType.FromString(value));
+        }
+
+        private static IEnumerable<object[]> FromObject_TestData()
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs]: Worth having some tests for values boxed as `object`?

File: src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        [MemberData(nameof(FromString_TestData))]
+        public void FromObject(object value, byte expected)
+        {
+            Assert.Equal(expected, ByteType.FromObject(value));
+        }
+
+        // The following conversions are not supported.
+        [Theory]
+        [MemberData(nameof(FromObject_NotSupported_TestData))]
+        public void FromObject_NotSupported(object value, byte expected)
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs]: >, byte expected [](start = 56, length = 15)

Why does this have the extra param?

File: src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs (hunk 4 above)
Comment (333fred) [reply src/Microsoft.VisualBasic.Core/tests/ByteTypeTests.cs]: Nevermind, I'm an idiot

---
In reply to: [277031493](https://github.com/dotnet/corefx/pull/37010#discussion_r277031493) [](ancestors = 277031493)

File: src/Microsoft.VisualBasic.Core/tests/DateTypeTests.cs [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        private static IEnumerable<object[]> FromObject_Overflow_TestData()
+        {
+            yield break;
+        }
+
+        private static IEnumerable<object[]> FromString_TestData()
+        {
+            yield return new object[] { "12:00:00 AM", new DateTime(0) };
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/tests/DateTypeTests.cs]: Consider having one more test to verify that it doesn't always return `default`.

**Comment by danmosemsft** [comment]
I just realized these tests aren't running against NETFX, although that's important for the reason above. This diff will do the trick I think:
```
diff --git a/src/Microsoft.VisualBasic.Core/tests/Configurations.props b/src/Microsoft.VisualBasic.Core/tests/Configurations.props
index acf56fa275..0ea07aa8f9 100644
--- a/src/Microsoft.VisualBasic.Core/tests/Configurations.props
+++ b/src/Microsoft.VisualBasic.Core/tests/Configurations.props
@@ -3,6 +3,7 @@
     <BuildConfigurations>
       netcoreapp;
       uap;
+      netfx;
     </BuildConfigurations>
   </PropertyGroup>
 </Project>
\ No newline at end of file
diff --git a/src/Microsoft.VisualBasic.Core/tests/VB/Configurations.props b/src/Microsoft.VisualBasic.Core/tests/VB/Configurations.props
index f97cd5ea7a..079a372338 100644
--- a/src/Microsoft.VisualBasic.Core/tests/VB/Configurations.props
+++ b/src/Microsoft.VisualBasic.Core/tests/VB/Configurations.props
@@ -3,6 +3,7 @@
     <BuildConfigurations>
       netcoreapp;
       uap;
+      netfx;
     </BuildConfigurations>
   </PropertyGroup>
 </Project>
```

After that I ran `dotnet test /p:targetgroup=netfx ` on each of the test projects.

On the primary (C#) one I got one failure which is not too bad:
```
      <test name="Microsoft.VisualBasic.Tests.ErrObjectTests.FilterDefaultMessage" type="Microsoft.VisualBasic.Tests.ErrObjectTests" method="FilterDefaultMessage" time="0.0070903" result="Fail">
        <failure exception-type="Xunit.Sdk.EqualException">
          <message><![CDATA[Assert.Equal() Failure\r\n          ↓ (pos 0)\r\nExpected: ID53\r\nActual:   File not found.\r\n          ↑ (pos 0)]]></message>
          <stack-trace><![CDATA[   at Microsoft.VisualBasic.Tests.ErrObjectTests.FilterDefaultMessage() in c:\git\corefx\src\Microsoft.VisualBasic.Core\tests\ErrObjectTests.cs:line 54]]></stack-trace>
        </failure>
      </test>
```
On the VB one I got a compile error that will need investigation
```
c:\git\corefx\src\Microsoft.VisualBasic.Core\tests\VB\FileIOTestBase.vb(12,18): error BC30002: Type 'IO.FileCleanupTestBase' is not defined. [c:\git\corefx\src\Microsoft.VisualBasic.Core\tests\VB\Microsoft.VisualBasic.VB.Core.Tests.vbproj]
```

@cston let me know if you have questions.

**Comment by agocke** [comment]
@cston It's hard to  review this without the context of where it came from. Is most of this copied verbatim? If it's not, can you summarize where you made changes and how you tested the changes?

**Comment by 333fred** [comment]
@agocke sources come from referencesource: https://github.com/Microsoft/referencesource/tree/master/Microsoft.VisualBasic

**Comment by cston** [comment]
@agocke, the source was mostly copied verbatim. There were minor cosmetic changes to comments and assert text, but the only changes to the implementation were:
- `LikeOperator.ExpandString()` does not handle full/half width mapping
- `VBBinder.InvokeMember()` does not handle `objType.IsCOMObject()`
- `LateBinding.GetCorrectIReflect()` does not handle `IsCOMObject()`.

**Comment by danmosemsft** [comment]
@agocke I don't know about these tests, but FYI the .NET Framework tests are visible at https://github.com/dotnet/corefx/compare/master...danmosemsft:old.vb.tests?expand=1

**Comment by cston** [comment]
The tests added in this PR for the various `Type` classes are based on similar tests from `ConversionsTests.vb`. The `LikeOperator` and `LateBinding` tests are new.

**Comment by cston** [comment]
@danmosemsft, I've logged https://github.com/dotnet/corefx/issues/37037 for the compile error. I've verified there are no new test failures with .NET Framework, and I'll address the existing failure with `ErrObjectTests.FilterDefaultMessage` in a subsequent PR.

**Comment by danmosemsft** [comment]
sounds good to me!
